File: packages/sanjaya-django/src/sanjaya_django/schemas/ssrm.py
"""Shared AG Grid SSRM schemas used by both the table and pivot endpoints."""

from __future__ import annotations

from typing import Any

from sanjaya_core.enums import AggFunc, ColumnType

from sanjaya_django.schemas import CamelSchema


# ---------------------------------------------------------------------------
# Shared types (used by table, pivot, and export)
# ---------------------------------------------------------------------------


class ColumnVO(CamelSchema):
    id: str
    display_name: str
    field: str | None = None
    agg_func: AggFunc | None = None


class SortModelItem(CamelSchema):
    col_id: str
    sort: str  # "asc" | "desc"


# ---------------------------------------------------------------------------
# Shared request base
# ---------------------------------------------------------------------------


class SSRMBaseRequest(CamelSchema):
    """Fields common to both the table and pivot SSRM endpoints."""

    start_row: int
    end_row: int
    row_group_cols: list[ColumnVO]
    group_keys: list[str]
    value_cols: list[ColumnVO]
    sort_model: list[SortModelItem] | None = None
    filter: dict[str, Any] | None = None
    filter_model: dict[str, Any] | None = None


# ---------------------------------------------------------------------------
# Table request (no pivot)
# ---------------------------------------------------------------------------


class TableGetRowsRequest(SSRMBaseRequest):
    """AG Grid SSRM request for the table endpoint.

    Supports flat data, row grouping, and aggregation — but no pivoting.
    """


# ---------------------------------------------------------------------------
# Response (shared by both endpoints)
# ---------------------------------------------------------------------------


class PivotResultColDef(CamelSchema):
    col_id: str
    header_name: str
    field: str
    type: ColumnType | None = None
    children: list[PivotResultColDef] | None = None
    pivot_meta: dict | None = None


class ServerSideGetRowsResponse(CamelSchema):
    row_data: list[dict[str, Any]]
    row_count: int | None = None
    pivot_result_fields: list[str] | None = None
    secondary_col_defs: list[PivotResultColDef] | None = None

File: packages/sanjaya-ui/src/types/ssrm.ts
// ---------------------------------------------------------------------------
// AG Grid SSRM types — mirror sanjaya_django.schemas.ssrm / pivot / export
// ---------------------------------------------------------------------------

import type { ColumnType } from "./columns.js";

// ---------------------------------------------------------------------------
// Aggregate function enum
// ---------------------------------------------------------------------------

/**
 * Aggregate functions for pivot measures.
 *
 * Values align with AG Grid `aggFunc` strings.
 */
export enum AggFunc {
  Sum = "sum",
  Avg = "avg",
  Min = "min",
  Max = "max",
  Count = "count",
  DistinctCount = "distinctCount",
  First = "first",
  Last = "last",
}

// ---------------------------------------------------------------------------
// Shared column / sort descriptors
// ---------------------------------------------------------------------------

/**
 * AG Grid column descriptor sent in SSRM requests.
 *
 * Used for `rowGroupCols`, `pivotCols`, and `valueCols`.
 */
export interface ColumnVO {
  id: string;
  displayName: string;
  field?: string;
  aggFunc?: AggFunc;
}

/**
 * AG Grid sort model item.
 */
export interface SortModelItem {
  colId: string;
  sort: "asc" | "desc";
}

// ---------------------------------------------------------------------------
// Requests
// ---------------------------------------------------------------------------

/**
 * Fields common to both table and pivot SSRM requests.
 */
export interface SSRMBaseRequest {
  startRow: number;
  endRow: number;
  rowGroupCols: ColumnVO[];
  groupKeys: string[];
  valueCols: ColumnVO[];
  sortModel?: SortModelItem[];
  /** Rich FilterGroup JSON (takes precedence over filterModel). */
  filter?: Record<string, unknown>;
  /** AG Grid column-keyed filter model (legacy / fallback). */
  filterModel?: Record<string, unknown>;
}

/**
 * SSRM request for the table endpoint (flat / grouped, no pivot).
 */
export interface TableSSRMRequest extends SSRMBaseRequest {
  // No additional fields — table never pivots.
}

/**
 * SSRM request for the pivot endpoint.
 */
export interface PivotSSRMRequest extends SSRMBaseRequest {
  pivotCols: ColumnVO[];
  pivotMode: boolean;
}

// ---------------------------------------------------------------------------
// Response
// ---------------------------------------------------------------------------

/**
 * Pivot result column definition for programmatic header control.
 */
export interface PivotResultColDef {
  colId: string;
  headerName: string;
  field: string;
  type?: ColumnType;
  children?: PivotResultColDef[];
  pivotMeta?: {
    pivotKeys?: string[];
    measure?: string;
    agg?: string;
  };
}

/**
 * SSRM response (shared by table and pivot endpoints).
 */
export interface SSRMResponse {
  rowData: Record<string, unknown>[];
  rowCount?: number;
  pivotResultFields?: string[];
  secondaryColDefs?: PivotResultColDef[];
}

// ---------------------------------------------------------------------------
// Export
// ---------------------------------------------------------------------------

/**
 * Supported export output formats.
 */
export enum ExportFormat {
  Csv = "csv",
  Xlsx = "xlsx",
}

export interface FlatExportRequest {
  selectedColumns: string[];
  filter?: Record<string, unknown>;
  format: ExportFormat;
}

export interface GroupedExportRequest {
  rowGroupCols: ColumnVO[];
  valueCols: ColumnVO[];
  filterModel?: Record<string, unknown>;
  sortModel?: SortModelItem[];
  format: ExportFormat;
}

export interface PivotExportRequest {
  rowGroupCols: ColumnVO[];
  valueCols: ColumnVO[];
  pivotCols: ColumnVO[];
  filterModel?: Record<string, unknown>;
  sortModel?: SortModelItem[];
  format: ExportFormat;
}

/**
 * Export request — exactly one of `flat`, `grouped`, or `pivot` must be set.
 */
export interface ExportRequest {
  flat?: FlatExportRequest;
  grouped?: GroupedExportRequest;
  pivot?: PivotExportRequest;
}
